docs/proposals/dns-using-envoy.md

```diff
@@ -0,0 +1,73 @@
+# Use Envoy DNSFilter for DNS resolution
+
+## Context
+
+Currently, Kuma relies on an external DNS to provide VIPs which are then added as services which proxy back to the actual Cluster in Envoy.
+For VIPs to be resolved Kuma CP runs a DNS server which maintains this mapping.
+
+This current setup has some limitations:
+
+- Apps need to set their DNS to point to the CP's DNS (or CNIs need to configure DNS to be correct)
+- The CP needs to run a DNS and handle the load
+- When the CP is down DNS requests will fail. This is usually not the case in a Service Mesh
+- DNS configuration is global to all meshes
+- DNS is a separate component which makes it less customisable compared to things that end up in XDS
+- This DNS is not aware of the origin of requests, therefore isolation of meshes breaks as anyone can resolve any domain
+
+In this proposal we suggest using [Envoy's DNSFilter](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/udp_filters/dns_filter) to replace the CP DNS.
+We add a new `DNS policy` to be able to control the configuration of the DNS.
+
+## Out of scope
+
+* UDP transparent proxying
+* Handling SRV records. While DNSFilter supports it, it's probably not necessary in v1
+
+## Requirements
+
+* The top-level domain should be configurable (defaults to `mesh`)
+* Drop in replacement to Kuma DNS (Default configuration should match Kuma's current configuration)
+* ability to set upstream dns resolvers to have a single resolver in the configuration
+* Select the tag(s) to use for lower level domain names
+* Provide complete isolation between meshes. .i.e. it shouldn't be possible from a mesh to figure out services running in other meshes by querying the DNS
+* DNS query should still succeed even if the CP is down
+
+## Configuration model
+
+Here is a full example of the configuration
+
+```yaml
+type: DNS
+mesh: default
+name: dns-1
+selectors:
+  - match:
+      kuam.io/service: "*"
+spec:
+  domain: "kuma" # [Optional defaults to "mesh"] the TLD to use for services exposed with the DNS
+  upstreamResolvers: # [Optional] A list of resolvers to use when we resolving names not in the TLD
+    - "8.8.8.8"
+    - "4.4.4.4"
```

Comment:
Ahem, in K8s we'll have to add here also whatever ends up in /etc/resolv.conf. We do not want to break the normal K8s service resolving, i.e. resolving to Cluster IPs

Reply:
I was also expecting that our Envoy will just do the resolving of Kuma domains.

Do you need this option @lahabana for your deployment?
If not, I'd consider not including this for now

Reply:
No we don't I'll remove it then.